profile/encode.go

```diff
@@ -288,15 +292,31 @@ func (p *Profile) postDecode() error {
 
 	for _, s := range p.Sample {
 		labels := make(map[string][]string, len(s.labelX))
-		numLabels := make(map[string][]int64, len(s.labelX))
+		numLabels := make(map[string]NumValues, len(s.labelX))
 		for _, l := range s.labelX {
 			var key, value string
 			key, err = getString(p.stringTable, &l.keyX, err)
 			if l.strX != 0 {
 				value, err = getString(p.stringTable, &l.strX, err)
 				labels[key] = append(labels[key], value)
 			} else if l.numX != 0 {
-				numLabels[key] = append(numLabels[key], l.numX)
+				var unit string
+				if l.unitX != 0 {
+					unit, err = getString(p.stringTable, &l.unitX, err)
+				} else if key == "bytes" || key == "requests" || key == "alignment" {
+					unit = "bytes"
```

Comment:
I don't think it is right for pprof to force "bytes" as the units when writing profiles. I think we should preserve whatever was in the incoming profile.

It is OK for pprof to internally guess the units based on heuristics, but I would prefer to not expose those outside pprof.

Reply:
I could see guessing the units everywhere that tag units are used getting a bit tricky, but I can look more into it if that seems like the best option.

Would it work to add a field to indicate that units have been guessed and then not write out the tag units if this field indicates that units were guessed?

Reply:
I'd like to at least explore how bad it is. I'd like to keep this as a self-contained hack that we can eventually easily remove.

In particular I worry that if somebody has in the future a profile with tags {key =requests, unit=count}, we'd override their unit to bytes incorrectly.

Reply:
Right now, this won't ever override units that were already assigned -- just add units when the units are empty.  (I'll update the tests to capture this)
It is true, if someone specified the units as "", then this would infer incorrect units.

Reply:
Yes, I think it is OK for pprof to infer incorrect units in some cases, but I'm weary of baking units derived from heuristics into profiles we generate.

Reply:
I've modified this so inferred units will not modify the profile (if the profile is, for example, saved)